### PR TITLE
Use noCache when cache disabled

### LINKED ISSUE #18749: Setting `cache.ENABLED=false` causes 500 panic
### Gitea Version

1.15.11

### Git Version

_No response_

### Operating System

Docker in Liunx

### How are you running Gitea?

Via docker (docker swarm):

```
  gitea-server:
    image: gitea/gitea:1.15
    volumes:
      - /data/volumes/shadoware.org/gitea:/data
    environment:
      - SOCAT_LINK=false
      - USER_UID=1000
      - USER_GID=1000
    networks:
      - sendmail
      - traefik-swarm
    deploy:
      labels:
        - traefik.enable=true
        - traefik.http.routers.shadoware_gitea_router.rule=Host(`gogs.shadoware.org`)
        - traefik.http.routers.shadoware_gitea_router.tls
        - traefik.http.routers.shadoware_gitea_router.tls.certresolver=shadoware
        - traefik.http.routers.shadoware_gitea_router.tls.domains[0].main=shadoware.org
        - traefik.http.routers.shadoware_gitea_router.tls.domains[0].sans=*.shadoware.org
        - traefik.http.services.shadoware_gitea.loadbalancer.server.port=3000
      restart_policy:
        condition: any
      update_config:
        order: start-first
      resources:
        limits:
          cpus: "1"
          memory: 512M
      placement:
        constraints:
          - node.labels.APPLICATION == true
```

the configuration is 

```
APP_NAME = Gogs: Go Git Service
RUN_USER = git
RUN_MODE = prod

[database]
DB_TYPE  = sqlite3
HOST     = 127.0.0.1:3306
NAME     = gogs
USER     = root
PASSWD   = 
SSL_MODE = disable
PATH     = /data/gitea/data/gitea.db

[repository]
ROOT = /data/git/gitea-repositories

[server]
DOMAIN       = gogs.shadoware.org
HTTP_PORT    = 3000
ROOT_URL     = https://gogs.shadoware.org/
DISABLE_SSH  = true
SSH_PORT     = 22
OFFLINE_MODE = false
LANDING_PAGE = explore
ENABLE_GZIP  = true
LFS_START_SERVER = true
APP_DATA_PATH = /data/gitea

[mailer]
ENABLED = true
HOST = posteio:25
FROM = noreply@shadoware.org

[service]
REGISTER_EMAIL_CONFIRM = false
ENABLE_NOTIFY_MAIL     = false
DISABLE_REGISTRATION   = true
ENABLE_CAPTCHA         = false
REQUIRE_SIGNIN_VIEW    = false

[picture]
DISABLE_GRAVATAR        = true
ENABLE_FEDERATED_AVATAR = false

[session]
PROVIDER = file
PROVIDER_CONFIG = /data/sessions

[cache]
ENABLED=false
ITEM_TTL=0

[log]
MODE      = file
LEVEL     = Warn
ROOT_PATH = /data/gitea/log

[log.console]
LEVEL = Warn

[log.file]
LEVEL = Warn

[security]
INSTALL_LOCK   = true
SECRET_KEY     = hidden
INTERNAL_TOKEN = hidden

[oauth2]
JWT_SECRET = hidden

[attachment]
ENABLED = true
PATH = /data/gitea/data/attachments
ALLOWED_TYPES = */*
MAX_SIZE = 50
MAX_FILES = 20

[lfs]
STORAGE_TYPE=local
PATH=/data/gitea/data/lfs

[storage.repo-archive]
STORAGE_TYPE=local
PATH=/data/gitea/data/repo-archive
```

### Database

SQLite

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Description

I have the following error when i go to the page "Profil / Configuration / Account":

(On IHM i have 500)

2022/02/13 15:03:58 routers/web/base.go:130:1() [E] PANIC: runtime error: invalid memory address or nil pointer dereference
        /usr/local/go/src/runtime/panic.go:212 (0x43ea3a)
        /usr/local/go/src/runtime/signal_unix.go:734 (0x458a12)
        /go/src/code.gitea.io/gitea/routers/web/user/setting/account.go:297 (0x21bdbb6)
        /go/src/code.gitea.io/gitea/routers/web/user/setting/account.go:35 (0x21b9764)
        /go/src/code.gitea.io/gitea/modules/web/route.go:64 (0x2038a9b)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/vendor/github.com/go-chi/chi/mux.go:436 (0x16f9f0a)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/modules/web/route.go:103 (0x2039176)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/modules/web/route.go:103 (0x2039176)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/modules/web/route.go:103 (0x2039176)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/vendor/github.com/go-chi/chi/middleware/get_head.go:37 (0x205b5c1)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/modules/web/route.go:103 (0x2039176)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/modules/context/context.go:791 (0x171ae49)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/vendor/github.com/NYTimes/gziphandler/gzip.go:338 (0x2268498)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/vendor/github.com/go-chi/chi/mux.go:70 (0x16f7a6a)
        /go/src/code.gitea.io/gitea/vendor/github.com/go-chi/chi/mux.go:311 (0x16fde5b)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/vendor/github.com/go-chi/chi/mux.go:436 (0x16f9f0a)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/routers/web/base.go:169 (0x22702f0)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/go-chi/session/session.go:257 (0x1506fce)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/modules/public/public.go:42 (0x167bc8e)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/vendor/github.com/go-chi/chi/mux.go:70 (0x16f7a6a)
        /go/src/code.gitea.io/gitea/vendor/github.com/go-chi/chi/mux.go:311 (0x16fde5b)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/vendor/github.com/go-chi/chi/mux.go:436 (0x16f9f0a)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/routers/common/middleware.go:75 (0x206005d)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/vendor/github.com/go-chi/chi/middleware/strip.go:30 (0x205bec7)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/vendor/github.com/chi-middleware/proxy/middleware.go:37 (0x205752e)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/routers/common/middleware.go:28 (0x205fba5)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /go/src/code.gitea.io/gitea/vendor/github.com/go-chi/chi/mux.go:87 (0x16f77f0)
        /go/src/code.gitea.io/gitea/modules/web/route.go:318 (0x2037fd3)
        /go/src/code.gitea.io/gitea/vendor/github.com/gorilla/context/context.go:141 (0x1212813)
        /usr/local/go/src/net/http/server.go:2050 (0x7c2763)
        /usr/local/go/src/net/http/server.go:2868 (0x7c5d22)
        /usr/local/go/src/net/http/server.go:1933 (0x7c124c)
        /usr/local/go/src/runtime/asm_amd64.s:1371 (0x47b360)

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by lunny**
close #19703
fix #18749

### CODE REVIEW

**Review by 6543 — APPROVED** [review]

**Comment by wxiaoguang** [comment]
The no-op cache will make the email-resend-limit lose the protection. If some users disable the cache, they might be exposed to the email spam risk.

https://github.com/go-gitea/gitea/pull/19703#issuecomment-1126851883

**Comment by 6543** [comment]
@wxiaoguang if we do depend on cache already why do we have an option to disable it at all?

**Comment by wxiaoguang** [comment]
> @wxiaoguang if we do depend on cache already why do we have an option to disable it at all?

When I joined this project, the option was already there. That's why I proposed to remove the option to disable it, and always use a builtin cache.

In fact, that's how a big project evolves. At the beginning, an option is optional. When new code comes, some optional options changes, they may become required. Such changes, some are caused by well-designed purpose, some are caused by not-well-designed features.

Unfortunately, the  email-resend-limit feature was not well designed, to be honest, it did take the consideration that the cache could be disabled, then the app panics.

File: modules/cache/empty.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+func (c noCache) Put(key string, val interface{}, timeout int64) error {
+	return nil
+}
+
+// Get gets cached value by given key.
+func (c noCache) Get(key string) interface{} {
+	// workaround until https://gitea.com/go-chi/cache/issues/3 is resolved
+	if key == pingTestKey {
+		return pingTestVal
+	}
```
Comment (6543) [inline modules/cache/empty.go]:
```suggestion
```

**Review by zeripath — APPROVED** [review]

File: modules/cache/cache.go [hunk 2]
```diff
@@ -18,6 +18,10 @@ import (
 var conn mc.Cache
 
 func newCache(cacheConfig setting.Cache) (mc.Cache, error) {
+	if !cacheConfig.Enabled {
+		return newNoCache()
+	}
+
```
Comment (zeripath) [inline modules/cache/cache.go:24]: This breaks the unit tests.

**Comment by 6543** [comment]
#19703 can be packported ... this will need more code moving ... - I considder it a refactor

**Pull request closed without merging**